fl4health/utils/data_generation.py

```diff
@@ -0,0 +1,295 @@
+                encoding labels. Default is as in the FedProx paper. Defaults to 10.
+            samples_per_client (int, optional): Number of samples to generate in each client's dataset.
+                Defaults to 1000.
+        """
+        self.num_clients = num_clients
+        self.temperature = temperature
+        self.input_dim = input_dim
+        self.output_dim = output_dim
+        self.samples_per_client = samples_per_client
+
+        # Sigma in the FedProx paper
+        self.input_covariance = self.construct_covariance_matrix()
+
+    def construct_covariance_matrix(self) -> torch.Tensor:
+        """
+        This function generations the covariance matrix used in generating input features. It is fixed across all
+        datasets. It is a diagonal matrix with diagonal entries x_{j, j} = j^{-1.2}, where j starts at 1 in this
+        notation. The matrix is of dimension input_dim x input_dim
+
+        Returns:
+            torch.Tensor: Covariance matrix for generation of input features.
+        """
+        sigma_diagonal = torch.zeros(self.input_dim)
+        for i in range(self.input_dim):
+            # indexing in the original implementation starts at 1, so i+1
+            sigma_diagonal[i] = math.pow((i + 1), -1.2)
+        return torch.diag(sigma_diagonal)
+
+    def map_inputs_to_outputs(self, x: torch.Tensor, W: torch.Tensor, b: torch.Tensor) -> torch.Tensor:
+        """
+        This function maps features x to a label y as done in the original paper. The first stage is the affine
+        transformation hat{y} = (1/T)*(Wx + b). Then y = softmax(hat{y}). Sampling from the distribution, we then
+        one hot encode the resulting label sample.
+
+        NOTE: This procedure differs slightly from that of the original paper, which simply took a one hot on the
+        softmax distribution. The current strategy allows for a bit more label stochasticity.
+
+        Args:
+            x (torch.Tensor): The input features to be mapped to output labels. Shape is (dataset size, input_dim)
+            W (torch.Tensor): The linear transformation matrix. Shape is (output_dim, input_dim)
+            b (torch.Tensor): The bias in the linear transformation. Shape is (output_dim, 1)
+
+        Returns:
+            torch.Tensor: The labels associated with each of the inputs. The shape is (dataset size, output_dim)
+        """
+        raw_y = (torch.matmul(x, W.T) + b.T.repeat(self.samples_per_client, 1)) / self.temperature
+        distributions = F.softmax(raw_y, dim=1)
+        samples = torch.multinomial(distributions, 1)
+        return F.one_hot(samples, num_classes=self.output_dim).squeeze()
+
+    def generate(self) -> List[TensorDataset]:
+        """
+        Based on the class parameters, generate a list of synthetic TensorDatasets, one for each client.
+
+        Returns:
+            List[TensorDataset]: Synthetic datasets for each client.
+        """
+        client_tensors = self.generate_client_tensors()
```

Comment:
Super minor, but maybe it is worthwhile to assert that `generate_client_tensors` returns a list with length self.num_clients.

Reply:
Will add!